diff/autofd/autofd.go

```diff
@@ -0,0 +1,298 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package autofd
+
+import (
+	"fmt"
+	"go/ast"
+	"go/token"
+	"go/types"
+	"io"
+	"strings"
+
+	"golang.org/x/tools/go/packages"
+)
+
+type Func struct {
+	Path  string // import path of the package holding the function
+	Name  string // function or method name
+	Deriv string // name of the output derivative function
+}
+
+// Derivative generates the derivative code from the given function declaration.
+func Derivative(w io.Writer, f Func) error {
+	gen, err := newGenerator(w, f)
+	if err != nil {
+		return fmt.Errorf("could not create derivative generator: %w", err)
+	}
+	err = gen.generate()
+	if err != nil {
+		return fmt.Errorf("could not generate derivative: %w", err)
+	}
+	return nil
+}
+
+type generator struct {
+	w   io.Writer
+	pkg *packages.Package
+	fct *types.Func
+	der string
+	err error
+}
+
+func newGenerator(w io.Writer, f Func) (*generator, error) {
+	path := f.Path
+	name := f.Name
+
+	cfg := &packages.Config{
+		Mode: packages.NeedName |
+			packages.NeedFiles |
+			packages.NeedCompiledGoFiles |
+			packages.NeedSyntax |
+			packages.NeedTypes |
+			packages.NeedTypesInfo,
+	}
+	pkgs, err := packages.Load(cfg, path)
+	if err != nil {
+		return nil, fmt.Errorf("could not load package of %q: %w", f, err)
```

Comment:
I'd destructure `f` more for this; `("could not load package of %q %s: %w", f.Path, f.Name, err)`.